File: typescript/packages/mechanisms/cardano/src/exact/script/datum.ts
import { Data, InlineDatum } from "@evolution-sdk/evolution";

import type { CardanoExtraScript } from "../../types";
import { MAX_CARDANO_DATUM_BYTES } from "../../limits";

/**
 * Builds the inline datum a script payment attaches to its `payTo` output from
 * the script `extra`. The datum is the server-supplied CBOR hex
 * (`extra.datum`), decoded and wrapped as an inline datum so the funds land at
 * the contract with the datum it expects.
 *
 * Unlike the Masumi lock datum (a fixed 19-field escrow shape this package
 * builds), a script datum is arbitrary and contract-specific — the server that
 * defined the contract supplies it verbatim and owns its correctness. This
 * helper only parses it; the facilitator does not verify its contents.
 *
 * @param extra - The script `extra` block from the payment requirements.
 * @returns The inline datum for the `payTo` output, or `undefined` when the
 *   script needs no datum (`extra.datum` omitted).
 * @throws When `extra.datum` is present but not non-empty CBOR hex.
 */
export function buildScriptDatumInline(
  extra: CardanoExtraScript,
): InlineDatum.InlineDatum | undefined {
  if (extra.datum === undefined) {
    return undefined;
  }
  if (
    typeof extra.datum !== "string" ||
    extra.datum.length === 0 ||
    extra.datum.length % 2 !== 0 ||
    extra.datum.length / 2 > MAX_CARDANO_DATUM_BYTES
  ) {
    throw new Error('Cardano script payment "datum" must be non-empty CBOR hex');
  }
  let data: Data.Data;
  try {
    data = Data.fromCBORHex(extra.datum);
  } catch (cause) {
    throw new Error('Cardano script payment "datum" is not valid CBOR hex', { cause });
  }
  return new InlineDatum.InlineDatum({ data });
}

File: python/x402/mechanisms/cardano/exact/script/datum.py
"""Decode a bounded inline Plutus datum."""

import re
from typing import Any

from pycardano import RawPlutusData

from ...limits import MAX_CARDANO_DATUM_BYTES
from ...utils import decode_cbor


def build_script_datum_inline(extra: dict[str, Any]) -> RawPlutusData | None:
    """Return the supplied inline datum, if present."""
    if "datum" not in extra:
        return None
    value = extra["datum"]
    if (
        not isinstance(value, str)
        or not re.fullmatch(r"(?:[0-9a-fA-F]{2})+", value)
        or len(value) // 2 > MAX_CARDANO_DATUM_BYTES
    ):
        raise ValueError('Cardano script payment "datum" must be non-empty CBOR hex')
    try:
        raw = bytes.fromhex(value)
        decode_cbor(raw)
        return RawPlutusData.from_cbor(raw)
    except Exception as exc:
        raise ValueError('Cardano script payment "datum" is not valid Plutus data') from exc
